### PR TITLE
Cult 3.0

### PR DESCRIPTION
**Pull request by DeityLink**
:cl:
* experiment: Cult 3.0 is finally here, replacing Legacy Cult. The changelog for that update is so big that it's better that you think of it as a brand new game mode for now. The wiki will be updated in a few days. Do not hesitate to give feedback either on Github, on Discord, or in the thread.
* bugfix: Fixed a few vaults and away mission broken by old cult structures.

### CODE REVIEW

File: code/datums/gamemode/factions/bloodcult/bloodcult.dm [hunk 1]
```diff
@@ -53,15 +115,132 @@ var/veil_thickness = CULT_EPILOGUE//REMEMBER TO SET IT BACK TO CULT_PROLOGUE WHE
...
-			return TRUE
-			*/
-	return FALSE
+/datum/faction/bloodcult/process()
+	..()
+	if (change_cooldown)
+		change_cooldown--
+		if (change_cooldown <= 0)
+			var/datum/objective/bloodcult_sacrifice/O = locate() in objective_holder.objectives
+			if (!O.IsFulfilled())
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/bloodcult/bloodcult.dm:126]: Check if O exists first.

File: code/modules/html_interface/map/station_map.dm [hunk 2]
```diff
@@ -1,5 +1,5 @@
 #define HOLOMAP_OBSTACLE	"#FFFFFFDD"
-#define HOLOMAP_PATH		"#66666699"
+#define HOLOMAP_PATH		"#	"
```
Comment (DamianX) [inline code/modules/html_interface/map/station_map.dm:2]: ?

Comment (jknpj) [reply code/modules/html_interface/map/station_map.dm:2]: 🤔

File: maps/RandomZLevels/unused/backup/Academy.dmm [hunk 3]
```diff
@@ -0,0 +1,694 @@
+"aa" = (/turf/space,/area)
```
Comment (DamianX) [inline maps/RandomZLevels/unused/backup/Academy.dmm:1]: I don't think this file is supposed to exist

File: maps/randomvaults/backup/asteroid_temple.dmm [hunk 4]
```diff
@@ -0,0 +1,100 @@
+"aa" = (/turf/space,/area)
```
Comment (DamianX) [inline maps/randomvaults/backup/asteroid_temple.dmm:1]: I don't think this file is supposed to exist

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 5]
```diff
@@ -768,3 +1047,420 @@ var/list/cult_spires = list()
...
+
+/obj/structure/cult/pillar/New()
+	..()
+	var/turf/T = loc
+	for (var/obj/O in loc)
+		if (O != src)
+			O.ex_act(2)
+	T.ChangeTurf(/turf/simulated/floor/engine/cult)
+	T.turf_animation('icons/effects/effects.dmi',"cultfloor", 0, 0, MOB_LAYER-1, anim_plane = TURF_PLANE)
+	for (var/obj/structure/S in T)
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm:1073]: Surely the above obj itteration would take care of the subtypes such as structures and machinery.

File: code/datums/gamemode/role/cultist.dm [hunk 6]
```diff
@@ -73,6 +84,13 @@
 	to_chat(antag.current, "<span class='info'><a HREF='?src=\ref[antag.current];getwiki=[wikiroute]'>(Wiki Guide)</a></span>")
 	to_chat(antag.current, "<span class='sinister'>You find yourself to be well-versed in the runic alphabet of the cult.</span>")
 
+	spawn(1)
+		if (faction)
+			var/datum/objective_holder/OH = faction.objective_holder
+			var/datum/objective/O = null
+			for (var/datum/objective/OB in OH.objectives)
```
Comment (MadmanMartian) [inline code/datums/gamemode/role/cultist.dm:91]: Why not just get the last entry in the objectives list via OB = OH.objectives[OH.objectives.len]?

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 7]
```diff
@@ -280,60 +338,276 @@
...
+					update_icon()
+					overlays += "altar-soulstone3"
+					sleep (gem_delay/3)
+					altar_task = ALTARTASK_NONE
+					update_icon()
+					var/obj/item/device/soulstone/gem/gem = new (loc)
+					gem.pixel_y = 4
+
+/obj/structure/cult/altar/noncultist_act(var/mob/user)//Non-cultists can still remove blades planted on altars.
+	if(iscultist(user))
```
Comment (ShiftyRail) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm:505]: :thinking: this is noncultist_act, which is specifically called if `iscultist` failed

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 8]
```diff
@@ -280,60 +338,276 @@
...
+	else if (blade)
+		blade.forceMove(loc)
+		blade.attack_hand(user)
+		to_chat(user, "You remove \the [blade] from \the [src]</span>")
+		blade = null
+		playsound(loc, 'sound/weapons/blade1.ogg', 50, 1)
+		update_icon()
+		return
+	else
+		to_chat(user,"<span class='sinister'>You feel madness taking its toll, trying to figure out \the [name]'s purpose</span>")
```
Comment (ShiftyRail) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm:529]: A dot at the end (and in fact, at the end of most of the sentences here

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 9]
```diff
@@ -280,60 +338,276 @@
...
-			cultist_act(usr,"commune")
-		if ("soulstone")
-			to_chat(usr,"TODO: SPAWN A SOULSTONE")
-		if ("sacrifice")
-			to_chat(usr,"TODO: SACRIFICE")
-		if ("soulblade")
-			to_chat(usr,"TODO: IMBUE SOULBLADE")
+	if(href_list["signup"])
+		var/mob/M = usr
+		if(!M || !isobserver(M))
```
Comment (ShiftyRail) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm:537]: `isobserver(M)` is an `istype` and hence check against `M` being null

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 10]
```diff
@@ -280,60 +338,276 @@
...
+			return
+		var/obj/item/weapon/melee/soulblade/blade = locate() in src
+		if (!blade || !istype(blade))
```
Comment (ShiftyRail) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm:540]: Just `istype` works

File: code/modules/html_interface/map/station_map.dm (hunk 2 above)
Comment (DeityLink) [reply code/modules/html_interface/map/station_map.dm:2]: oh god

Comment (DeityLink) [reply code/modules/html_interface/map/station_map.dm:2]: Surely that won't break anything

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 11]
```diff
@@ -280,60 +338,276 @@
...
+			to_chat(usr, "<span class='warning'>The Soul Blade was pulled off from \the [src]</span>")
+			return
+		if (blade.shade)
+			to_chat(usr, "<span class='warning'>Another shade was faster, and is currently possessing the blade.</span>")
+			return
+		var/mob/living/simple_animal/shade/shadeMob = new(blade)
+		shadeMob.status_flags |= GODMODE
+		shadeMob.canmove = 0
+		shadeMob.ckey = usr.ckey
```
Comment (ShiftyRail) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm:549]: I'd rather use a complete mind transfer proc. Manually ckey editing is iffy.

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 12]
```diff
@@ -280,60 +338,276 @@
...
+		shadeMob.cancel_camera()
+		shadeMob.give_blade_powers()
+		blade.dir = NORTH
+		blade.update_icon()
+		update_icon()
+		//Automatically makes them cultists
+		var/datum/role/cultist/newCultist = new
+		newCultist.AssignToRole(shadeMob.mind,1)
+		var/datum/faction/bloodcult/cult = find_active_faction_by_type(/datum/faction/bloodcult)
+		if (!cult)
```
Comment (ShiftyRail) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm:566]: Jésus Marie Joseph, is there is no cult by this time then at the very least it should runtime

**Review by MadmanMartian — COMMENTED** [review]
AAAAAAAAAAAAA WHO SPEEDMERGES WITHIN THE FIRST 6 MINUTES

File: vgstation13.dme [hunk 13]
```diff
@@ -71,7 +71,6 @@
 #include "code\hub.dm"
 #include "code\names.dm"
 #include "code\world.dm"
-#include "code\modules\organs\internal\heart\heart.dm"
```
Comment (ShiftyRail) [inline vgstation13.dme:74]: Aaaah

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm (hunk 9 above)
Comment (ShiftyRail) [reply code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm:537]: Also usr is always a mob this typecast has no reason to exist actually

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm (hunk 12 above)
Comment (DeityLink) [reply code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm:566]: In case of admin varedit bullshit.

**Comment by I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE** [comment]
If you didn't send stats somewhere or at the very least log when each rune type is made to see the usage statistics I'm gonna be real damn sad.

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm (hunk 11 above)
Comment (MadmanMartian) [reply code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm:549]: I can see why it was done. Needed a clean mind, so you don't end up with a vampire possessing your blade and citing its vampirism for murderating the cult.

**Comment by DeityLink** [comment]
Shit I knew I forgot something, I'll add them soon

**Comment by SonixApache** [comment]
did you know that rounds are supposed to end?

**Comment by Intigracy** [comment]
This really should not have been merged as quickly as it was.

**Comment by DeityLink** [comment]
@SonixApache The rounds should end 5 minutes after Nar-Sie rises.

**Comment by ForrestWick** [comment]
Hi, I'm looking to have this ported to another server with some changes to balance and objectives. I figure it's only fair I ask if the author of Cult v3 is cool with it

**Comment by SonixApache** [comment]
cute!

**Comment by DamianX** [comment]
gamer

**Comment by DeityLink** [comment]
sure

**Pull request merged**